### PR TITLE
create workflow ci

### PR DESCRIPTION
**Pull request by samgawthrop**
testing a workflow ci

### CODE REVIEW

**Comment by samgawthrop** [comment]
you ok if I do a squash-and-merge, or should I crush these down to a single commit first?

File: pyquat/__init__.py [hunk 1]
```diff
@@ -11,6 +11,14 @@
 QUAT_SMALL = 1e-8
 
 
+def touch():
+    """Touch all imports from cython to alleviate flake8."""
+    if valenti_q_acc == valenti_q_mag:
+        print(rotation_vector_to_matrix)
+    identity(valenti_q_mag)
+    raise Exception("Never run this")
```
Comment (translunar) [inline pyquat/__init__.py:19]: I'm not clear on what this does.

File: pyquat/plot.py [hunk 2]
```diff
@@ -1,38 +1,36 @@
...
-            ary[:,i] = q.to_unit_vector(rotate_axis)[:,0]
+            ary[:, i] = q.to_unit_vector(rotate_axis)[:, 0]
     elif len(q_ary.shape) == 2:
         ary = numpy.empty((3, q_ary.shape[1]))
         for i, q in enumerate(q_ary):
-            ary[:,i] = q[0].to_unit_vector(rotate_axis)[:,0]
+            ary[:, i] = q[0].to_unit_vector(rotate_axis)[:, 0]
     else:
-        raise InputError("expected 1- or 2-D array")
+        raise Exception("expected 1- or 2-D array")
```
Comment (translunar) [inline pyquat/plot.py:17]: This should be a `ValueError`.

File: pyquat/plot.py [hunk 3]
```diff
@@ -53,14 +50,19 @@ def plot(q_ary, t = None, rotate_axis='x', fig = None, axes = None, **kwargs):
         tc = (t - t.min()) / (t.max() - t.min())
         c = numpy.vstack((tc, numpy.zeros_like(tc), -tc + 1.0))
 
-        for i in range(1,ary.shape[1]):
-            axes.plot(ary[0,i-1:i+1], ary[1,i-1:i+1], ary[2,i-1:i+1], c = tuple(c[:,i]), **kwargs)
+        for i in range(1, ary.shape[1]):
+            axes.plot(ary[0, i - 1:i + 1],
+                      ary[1, i - 1:i + 1],
+                      ary[2, i - 1:i + 1],
+                      c=tuple(c[:, i]), **kwargs)
```
Comment (translunar) [inline pyquat/plot.py:57]: I don't love this type of change. It's going to make some code absolutely unreadable. I realize that expressions should be space-separated in general, but these are expressions within bracket notation.

File: pyquat/plot.py [hunk 4]
```diff
@@ -70,21 +72,21 @@ def plot_frame(q, r = numpy.zeros((3,1)), fig = None, axes = None, axis_size = 1
...
-    xh = numpy.zeros((3,2))
+    xh = numpy.zeros((3, 2))
     yh = numpy.zeros_like(xh)
     zh = numpy.zeros_like(xh)
-    xh[0,1] = axis_size
-    yh[1,1] = axis_size
-    zh[2,1] = axis_size
+    xh[0, 1] = axis_size
+    yh[1, 1] = axis_size
+    zh[2, 1] = axis_size
```
Comment (translunar) [inline pyquat/plot.py:80]: Same comment as with the array indexing. I don't think this change substantially improves readability.

File: pyquat/random.py [hunk 5]
```diff
@@ -55,11 +54,9 @@ def rand(
 
     """
     if axis is not None and angle is not None:
-        raise StandardError("expected non-fixed angle or non-fixed axis or both")
+        raise Exception("expected non-fixed angle or non-fixed axis or both")
```
Comment (translunar) [inline pyquat/random.py:57]:
```suggestion
        raise ValueError("expected non-fixed angle or non-fixed axis or both")
```

File: test/test_pyquat.py [hunk 6]
```diff
@@ -4,7 +4,8 @@
 import math
 import unittest
 
-from .context import pq
+# from .context import pq
+import pyquat as pq
```
Comment (translunar) [inline test/test_pyquat.py:8]: When I did it this way before, it was trying to import the installed version of pyquat instead of the local tree. This is why the `context.py` file exists. Are you able to confirm that it's loading the local tree now and in all Python 3.x?

File: test/test_pyquat.py [hunk 7]
```diff
@@ -257,9 +258,9 @@ def test_expm(self):
         B1    = 13/51.0
         dt    = 0.05
         w     = np.array([[0.03, 0.02, 0.01]]).T
-        J     = np.diag([200.0, 200.0, 100.0])
-        J_inv = linalg.inv(J)
-        wk1   = pq.wdot(w, J)
+        # J     = np.diag([200.0, 200.0, 100.0])
+        # J_inv = linalg.inv(J)
+        # wk1   = pq.wdot(w, J)
```
Comment (translunar) [inline test/test_pyquat.py:263]: I believe this was supposed to test code that is used by `expm` before testing `expm` itself.

File: test/test_wahba_esoq.py [hunk 8]
```diff
@@ -58,7 +60,6 @@ def test_davenport_eigenvalues(self):
                         [1.0, 0.0],
                         [0.0, 1.0]])
         B = pq_esoq.attitude_profile_matrix(obs = obs, ref = ref)
-        irot = pq_esoq.sequential_rotation(B)
```
Comment (translunar) [inline test/test_wahba_esoq.py:61]: This line is needed. It does an in-place rotation on `B`.

File: test/test_wahba_valenti.py [hunk 9]
```diff
@@ -19,7 +15,7 @@ def test_q_acc_points_down(self):
             a1 = q.rotate(d)
             np.testing.assert_almost_equal(a0, a1, decimal=13)
 
-            q  = pqv.dq_acc(a0)
+            q  = pqv.q_acc(a0)
```
Comment (translunar) [inline test/test_wahba_valenti.py:18]: IIRC, these are different functions.

File: test/test_wahba_valenti.py [hunk 10]
```diff
@@ -36,7 +32,7 @@ def test_q_mag_points_north(self):
             l  = qa.conjugated().rotate(b)
             qm = pqv.q_mag(l)
 
-            dqm = pqv.dq_mag(l)
+            dqm = pqv.q_mag(l)
```
Comment (translunar) [inline test/test_wahba_valenti.py:35]: Also different functions.

File: test/test_wahba_valenti.py [hunk 11]
```diff
@@ -47,7 +43,7 @@ def test_q_mag_points_north(self):
             self.assertEqual(dqm.y, 0.0)
         
             q_G_to_L  = qa * qm
-            q_L_to_G  = q_G_to_L.conjugated()
+            # q_L_to_G  = q_G_to_L.conjugated()
```
Comment (translunar) [inline test/test_wahba_valenti.py:46]:
```suggestion

```

**Review by translunar — CHANGES_REQUESTED** [review]

File: pyquat/__init__.py (hunk 1 above)
Comment (samgawthrop) [reply pyquat/__init__.py:19]: This is an abomination to use the things imported from the _pyquat. I regret it already, but I'm not sure how to do this any better. 

Unused imports are bad, but the unused imports here are kinda-sorta necessary to load the context for the other things downrange that use this __init__ as the conduit into the cython. 

Cython itself didn't have any particularly clear patterns on how to explicitly import without leaving unused imports, or having unclear context.

File: pyquat/plot.py (hunk 2 above)
Comment (samgawthrop) [reply pyquat/plot.py:17]: fair I couldn't find a deceleration for InputError. I'll change these. I too dislike base Exceptions.

File: test/test_pyquat.py (hunk 7 above)
Comment (samgawthrop) [reply test/test_pyquat.py:263]: Is it still? these variables are unused.

File: test/test_wahba_esoq.py (hunk 8 above)
Comment (samgawthrop) [reply test/test_wahba_esoq.py:61]: so it should just be `pq_esoq.sequential_rotation(B)`? the irot variable isn't used. Also curious why it passes tests with an essential rotation removed, but that's a differnt question.

File: test/test_wahba_valenti.py (hunk 9 above)
Comment (samgawthrop) [reply test/test_wahba_valenti.py:18]: where's dq_acc declared? I see valenti_q_acc imported through the base __init__ , which is laundered through wahabi.valenti to be q_acc. 

```
from pyquat import valenti_q_mag as q_mag
from pyquat import valenti_q_acc as q_acc
```
but I don't see dq used elsewhere in the code, is that renamed elsewhere in an import?

these are getting hard to follow.

File: pyquat/plot.py (hunk 4 above)
Comment (translunar) [reply pyquat/plot.py:80]: Okay, I'm going to push back on the second part of this one a bit more because it doesn't actually appear in pep8: https://stackoverflow.com/questions/31741418/pep8-space-after-a-comma/37064013#37064013

I think there should be spaces after commas *except* within array notation. The examples in pep8 do show spaces after commas, even though this isn't specifically enumerated; but they don't show multi-dimensional array indexing notation.

Thoughts?

File: test/test_wahba_esoq.py (hunk 8 above)
Comment (translunar) [reply test/test_wahba_esoq.py:61]: I suggest an additional line that verifies that `irot` is not None. `self.assertNotEqual(irot, None)` maybe?

File: test/test_wahba_valenti.py (hunk 9 above)
Comment (translunar) [reply test/test_wahba_valenti.py:18]: I think we settled this during our zoom meeting yesterday. Two additional import lines should be added (in addition to the two imports you gave in your comment):

```
from pyquat import valenti_dq_mag as dq_mag
from pyquat import valenti_dq_acc as dq_acc
```

and the above change should be reverted.
